roadmap/implementers-guide/src/pvf-prechecking.md

```diff
@@ -21,7 +21,17 @@ As a result of this issue we need a fairly hard guarantee that the PVFs of regis
 
 ## Solution
 
-The problem is solved by having a pre-checking process which is run when a new validation code is included in the chain. A new PVF can be added in two cases:
+The problem is solved by having a pre-checking process.
+
+### Pre-checking
+
+Pre-checking mostly consists of attempting to prepare (compile) the PVF WASM blob. We use more strict limits (e.g. timeouts) here compared to regular preparation for execution. This way errors during preparation later are likely unrelated to the PVF itself, as it already passed pre-checking. We can treat such errors as local node issues.
+
+We also have an additional step where we attempt to instantiate the WASM runtime without running it. This is unrelated to preparation so we don't time it, but it does help us catch more issues.
```

Comment:
- Can we now treat runtime construction errors as internal errors? I didn't do that because the PVF can still be un-instantiable on certain machines/configurations while making it through the pre-check vote, making attacks against those environments possible.

Reply:
Technically, I think we have to have nodes abstain from voting in that case, anything else would be unfair: They voted against the PVF in pre-checking, but later we slash them because they fail validating a candidate? Does not sound fair.

On the other hand if it passed pre-checking, we know there is a super majority being able to validate the candidate, so it should be "fine" if some nodes don't vote. 

I do think we have to time the instantiation - it could be possible to make this take a very long time? 

At the same time, I would assume in the honest case for it to be very quick, so it should be ok to make this part of the preparation timeout in pre-checking.

Reply:
> I do think we have to time the instantiation - it could be possible to make this take a very long time?
>
> At the same time, I would assume in the honest case for it to be very quick, so it should be ok to make this part of > the preparation timeout in pre-checking.

We do have the long host-side timeout, but yeah that's not ideal. Lumping it in with the preparation timeout makes it harder to reason about as we are no longer comparing the same operations in prechecking vs. preparation.

Best would be to have a separate timeout for instantiation, as well as for instantiation in execution where it would be more lenient. This way we actually make sure that if it passes in prechecking it should pass in execution. And having it in a separate thread here as well as in execution means we can do more fine-grained sandboxing later. Or am I overcomplicating it? I can also just do your suggestion for now and file an issue for later.

Reply:
Given that our pre-checking timeout is already way lower than the preparation timeout, it seems fine to include the instantiation - assuming we make sure the timeout is such that for honest PVFs it should not trigger. So indeed, I think we can keep it simple. It does not have to be comparable - it has to be "lower enough". Given that execution timeout is in the seconds, which includes instantiation I can hardly imagine how this simplification could be causing issues. It should be clearly documented though.

Reply:
Hmm, we need the compiled artifact on-disk to run instantiation, and by that point the timeout has been stopped. We need to implement https://github.com/paritytech/substrate/issues/13861. Then we could check that the artifact is instantiable before we actually write it to disk.

Reply:
We'd still publish VRFs but then no-show I think.  In principle a vote against the PVF could mean "no-show early" but afaik there is no reason forthat extra code complexity, probably you just announce your VRF and then abandon it.

Reply:
> Hmm, we need the compiled artifact on-disk to run instantiation, and by that point the timeout has been stopped. We need to implement [paritytech/substrate#13861](https://github.com/paritytech/substrate/issues/13861). Then we could check that the artifact is instantiable before we actually write it to disk.

Let's just make sure there is some timeout. The rest is: "What is easiest to implement" - it should not matter too much.

Reply:
@eskimor Done. Once we have sandboxing I think this will need its own thread. For now it should be good.